### PR TITLE
COOPR-520: Add mutual auth support

### PR DESCRIPTION
**Pull request by artemmakogon**
Please review

### CODE REVIEW

File: src/main/java/co/cask/http/KeyStoreLoadException.java [hunk 1]
```diff
@@ -0,0 +1,28 @@
...
+ * License for the specific language governing permissions and limitations under
+ * the License.
+ */
+
+package co.cask.http;
+
+/**
+ * This exception indicates one of problems with keyStore load.
+ */
+public class KeyStoreLoadException extends Exception {
```
Comment (chtyim) [inline src/main/java/co/cask/http/KeyStoreLoadException.java]: Why not just use IOException?

File: src/main/java/co/cask/http/SSLConfig.java [hunk 2]
```diff
@@ -0,0 +1,98 @@
...
+  private SSLConfig(File keyStore, String keyStorePassword,
+                    String certificatePassword, File trustKeyStore, String trustKeyStorePassword) {
+    this.keyStore = keyStore;
+    this.keyStorePassword = keyStorePassword;
+    this.certificatePassword = certificatePassword;
+    this.trustKeyStore = trustKeyStore;
+    this.trustKeyStorePassword = trustKeyStorePassword;
+  }
+
+  public File getKeyStore() {
```
Comment (chtyim) [inline src/main/java/co/cask/http/SSLConfig.java:43]: javadocs are needed for all public methods.

File: src/main/java/co/cask/http/SSLConfig.java [hunk 3]
```diff
@@ -0,0 +1,98 @@
...
+
+package co.cask.http;
+
+import java.io.File;
+
+/**
+ * A class that encapsulates SSLContext configuration.
+ */
+public class SSLConfig {
+  private File keyStore;
```
Comment (chtyim) [inline src/main/java/co/cask/http/SSLConfig.java]: All these fields should be final.

File: src/main/java/co/cask/http/SSLHandlerFactory.java [hunk 4]
```diff
@@ -17,28 +17,39 @@
...
+  private static final Logger LOG = LoggerFactory.getLogger(SSLHandlerFactory.class);
+
   private static final String protocol = "TLS";
   private final SSLContext serverContext;
+  private boolean needClientAuth;
 
-  public SSLHandlerFactory(File keyStore, String keyStorePassword, String certificatePassword) {
-    if (keyStore == null) {
+  public SSLHandlerFactory(SSLConfig sslConfig) {
+    if (sslConfig.getKeyStore() == null) {
```
Comment (chtyim) [inline src/main/java/co/cask/http/SSLHandlerFactory.java]: The check should be done in the `SSLConfig.Builder`.

File: src/main/java/co/cask/http/SSLHandlerFactory.java [hunk 5]
```diff
@@ -17,28 +17,39 @@
...
       throw new IllegalArgumentException("Certificate Path Not Configured");
     }
-    if (keyStorePassword == null) {
+    if (sslConfig.getKeyStorePassword() == null) {
```
Comment (chtyim) [inline src/main/java/co/cask/http/SSLHandlerFactory.java]: Same here.

File: src/main/java/co/cask/http/SSLHandlerFactory.java [hunk 6]
```diff
@@ -48,28 +59,63 @@ public SSLHandlerFactory(File keyStore, String keyStorePassword, String certific
...
+    } catch (CertificateException e) {
+      throw new KeyStoreLoadException(e);
+    } catch (NoSuchAlgorithmException e) {
+      throw new KeyStoreLoadException(e);
+    } catch (KeyStoreException e) {
+      throw new KeyStoreLoadException(e);
+    } catch (IOException e) {
+      throw new KeyStoreLoadException(e);
+    } finally {
+      if (inputStream != null) {
```
Comment (chtyim) [inline src/main/java/co/cask/http/SSLHandlerFactory.java]: Use double try block to make logic simplier.

``` java
try {
  InputStream is = new FileInputStream(keyStore);
  try {
    ...
  } finally {
    Closeables.closeQuietly(is);
  }
} catch (...) {
}
```

File: src/main/java/co/cask/http/SSLHandlerFactory.java [hunk 7]
```diff
@@ -48,28 +59,63 @@ public SSLHandlerFactory(File keyStore, String keyStorePassword, String certific
...
+    } catch (FileNotFoundException e) {
+      throw new KeyStoreLoadException(e);
+    } catch (CertificateException e) {
+      throw new KeyStoreLoadException(e);
+    } catch (NoSuchAlgorithmException e) {
+      throw new KeyStoreLoadException(e);
+    } catch (KeyStoreException e) {
+      throw new KeyStoreLoadException(e);
+    } catch (IOException e) {
+      throw new KeyStoreLoadException(e);
```
Comment (chtyim) [inline src/main/java/co/cask/http/SSLHandlerFactory.java]: Seems like you just want to wrap all the non-runtime exception. You can simply do:

``` java
try {
  ...
} catch (Exception ex) {
  if (ex instanceof RuntimeException) {
    throw ((RuntimeException) ex);
  }
  throw new IOException(ex);
}
```

File: src/main/java/co/cask/http/SSLHandlerFactory.java [hunk 8]
```diff
@@ -48,28 +59,63 @@ public SSLHandlerFactory(File keyStore, String keyStorePassword, String certific
...
+    return ks;
+  }
+
   public SslHandler create() {
     SSLEngine engine = serverContext.createSSLEngine();
+    engine.setNeedClientAuth(needClientAuth);
     engine.setUseClientMode(false);
     return new SslHandler(engine);
   }
+
```
Comment (chtyim) [inline src/main/java/co/cask/http/SSLHandlerFactory.java]: Remove extra empty line.

File: src/main/java/co/cask/http/SSLHandlerFactory.java [hunk 9]
```diff
@@ -48,28 +59,63 @@ public SSLHandlerFactory(File keyStore, String keyStorePassword, String certific
...
+        this.needClientAuth = true;
+        KeyStore tks = getKeyStore(sslConfig.getTrustKeyStore(), sslConfig.getTrustKeyStorePassword());
+        tmf = TrustManagerFactory.getInstance(algorithm);
+        tmf.init(tks);
+      }
 
-      // Initialize the SSLContext to work with our key managers.
       serverContext = SSLContext.getInstance(protocol);
-      serverContext.init(kmf.getKeyManagers(), null, null);
+      serverContext.init(kmf.getKeyManagers(), tmf != null ? tmf.getTrustManagers() : null, null);
```
Comment (chtyim) [inline src/main/java/co/cask/http/SSLHandlerFactory.java]: Better initialize the `TrustManager[]` inside the `if` clause above and use it here directly. E.g.

``` java
KeyManager[] keyManagers = kmf.getKeyManagers();
TrustManager[] trustManagers = null;
if (sslConfig.getTrustKeyStore() != null) {
  ...
  TrustManagerFactory factory = TrustManagerFactory.getInstance(algorithm);
  factory.init(tks);
  trustManagers = tmf.getTrustManager();
}
serverContext.init(keyManagers, trustManagers, null);
```

File: README.md [hunk 10]
```diff
@@ -157,7 +157,8 @@ Code Sample
     NettyHttpService service = NettyHttpService.builder()
                                .setPort(7777)
                                .addHttpHandlers(ImmutableList.of(new ApplicationHandler()))
-                               .enableSSL(new File("/path/to/keyStore.jks"), "keyStorePassword", "certificatePassowrd")
+                               .enableSSL(SSLConfig.builder(new File("/path/to/keyStore.jks", "keyStorePassword")
+                                          .setCertificatePassword("certificatePassowrd").build())
```
Comment (albertshau) [inline README.md]: you didn't introduce it, but might as well fix the typo here: certificatePassowrd -> certificatePassword

**Comment by artemmakogon** [comment]
@albertshau @chtyim fixed

File: src/main/java/co/cask/http/SSLHandlerFactory.java [hunk 11]
```diff
@@ -16,59 +16,86 @@
...
+        ks.load(is, keyStorePassword.toCharArray());
+      } catch (Exception ex) {
+        if (ex instanceof RuntimeException) {
+          throw ((RuntimeException) ex);
+        }
+        throw new IOException(ex);
+      } finally {
+        Closeables.closeQuietly(is);
+      }
+    } catch (FileNotFoundException e) {
```
Comment (chtyim) [inline src/main/java/co/cask/http/SSLHandlerFactory.java]: `FileNotFoundException` extends `IOException`, you don't need to catch and rethrow.

**Comment by chtyim** [comment]
Just one comment left. Otherwise LGTM.

**Comment by artemmakogon** [comment]
@chtyim fixed

**Pull request merged**